tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/firebase/CassandraFirebaseSubscriptionRepository.scala

```diff
@@ -0,0 +1,99 @@
+  override def configure(): Unit = {
+    bind(classOf[CassandraFirebaseSubscriptionRepository]).in(Scopes.SINGLETON)
+    bind(classOf[FirebaseSubscriptionRepository]).to(classOf[CassandraFirebaseSubscriptionRepository])
+
+    Multibinder.newSetBinder(binder, classOf[CassandraModule])
+      .addBinding().toInstance(CassandraFirebaseSubscriptionTable.MODULE)
+
+    bind(classOf[CassandraFirebaseSubscriptionDAO]).in(Scopes.SINGLETON)
+  }
+}
+
+class CassandraFirebaseSubscriptionRepository @Inject()(dao: CassandraFirebaseSubscriptionDAO, clock: Clock) extends FirebaseSubscriptionRepository {
+
+  override def save(username: Username, request: FirebaseSubscriptionCreationRequest): Publisher[FirebaseSubscription] = {
+    val firebaseSubscription = FirebaseSubscription.from(request, evaluateExpiresTime(OptionConverters.toJava(request.expires.map(_.value)), clock))
+
+    isDuplicatedDeviceClientId(username, request.deviceClientId.value)
+      .handle((isDuplicatedDeviceClientId: Boolean, sink: SynchronousSink[AnyRef]) => {
+        if (isInThePast(request.expires, clock)) {
+          sink.error(ExpireTimeInvalidException(request.expires.get.value, "expires must be greater than now"))
+        }
+        if (isDuplicatedDeviceClientId) {
+          sink.error(DeviceClientIdInvalidException(request.deviceClientId, "deviceClientId must be unique"))
+        }})
+      .`then`(isDuplicatedDeviceToken(username, request.token.value))
+      .handle((isDuplicatedDeviceToken: Boolean, sink: SynchronousSink[AnyRef]) => {
+        if (isDuplicatedDeviceToken) {
+          sink.error(DeviceTokenInvalidException("deviceToken must be unique"))
+        }})
+      .`then`(dao.insert(username, firebaseSubscription))
+      .`then`(SMono.just(firebaseSubscription))
+  }
+
+  override def updateExpireTime(username: Username, id: FirebaseSubscriptionId, newExpire: ZonedDateTime): Publisher[FirebaseSubscriptionExpiredTime] =
+    SMono.just(newExpire)
+      .handle((inputTime: ZonedDateTime, sink: SynchronousSink[AnyRef]) => {
+        if (newExpire.isBefore(ZonedDateTime.now(clock))) {
+          sink.error(ExpireTimeInvalidException(inputTime, "expires must be greater than now"))
+        }})
+      .`then`(retrieveByFirebaseSubscriptionId(username, id)
+        .flatMap((subscription: FirebaseSubscription) => dao.insert(username, subscription.withExpires(evaluateExpiresTime(Optional.of(newExpire), clock))))
+        .map(_.expires)
+        .switchIfEmpty(SMono.error(FirebaseSubscriptionNotFoundException(id))))
+
+  override def updateTypes(username: Username, id: FirebaseSubscriptionId, types: util.Set[TypeName]): Publisher[Void] =
+    retrieveByFirebaseSubscriptionId(username, id)
+      .map(_.withTypes(CollectionConverters.asScala(types).toSeq))
+      .flatMap(dao.insert(username, _))
+      .switchIfEmpty(SMono.error(FirebaseSubscriptionNotFoundException(id)))
+      .`then`()
```

Comment:
```suggestion
    retrieveByFirebaseSubscriptionId(username, id)
      .switchIfEmpty(SMono.error(FirebaseSubscriptionNotFoundException(id)))
      .map(_.withTypes(CollectionConverters.asScala(types).toSeq))
      .flatMap(dao.insert(username, _))
      .`then`()
```

is better? 
// I know it is not difference it this case, but sometime, `map` or `flatmap` can return empty, and it throw notFoundException, that is wrong logic